src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/LinkedEditingRange/LinkedEditingRangeEndpoint.cs

```diff
@@ -0,0 +1,133 @@
+    {
+        // The regex below excludes characters that can never be valid in a TagHelper name.
+        // This is loosely based off logic from the Razor compiler:
+        // https://github.com/dotnet/aspnetcore/blob/main/src/Razor/Microsoft.AspNetCore.Razor.Language/src/Legacy/HtmlTokenizer.cs
+        // Internal for testing only.
+        internal readonly string _wordPattern = @"!?[^ <>!\/\?\[\]=""\\@" + Environment.NewLine + "]+";
+
+        private readonly ProjectSnapshotManagerDispatcher _projectSnapshotManagerDispatcher;
+        private readonly DocumentResolver _documentResolver;
+        private readonly TagHelperFactsService _tagHelperFactsService;
+
+        public LinkedEditingRangeEndpoint(
+            ProjectSnapshotManagerDispatcher projectSnapshotManagerDispatcher,
+            DocumentResolver documentResolver,
+            TagHelperFactsService tagHelperFactsService)
+        {
+            if (projectSnapshotManagerDispatcher is null)
+            {
+                throw new ArgumentNullException(nameof(projectSnapshotManagerDispatcher));
+            }
+
+            if (documentResolver is null)
+            {
+                throw new ArgumentNullException(nameof(documentResolver));
+            }
+
+            if (tagHelperFactsService is null)
+            {
+                throw new ArgumentNullException(nameof(tagHelperFactsService));
+            }
+
+            _projectSnapshotManagerDispatcher = projectSnapshotManagerDispatcher;
+            _documentResolver = documentResolver;
+            _tagHelperFactsService = tagHelperFactsService;
+        }
+
+        public RegistrationExtensionResult GetRegistration()
+        {
+            const string AssociatedServerCapability = "linkedEditingRangeProvider";
+            var registrationOptions = new LinkedEditingRangeRegistrationOptions();
+            return new RegistrationExtensionResult(AssociatedServerCapability, registrationOptions);
+        }
+
+        public async Task<LinkedEditingRanges> Handle(
+            LinkedEditingRangeParams request,
+            CancellationToken cancellationToken)
+        {
+            var document = await _projectSnapshotManagerDispatcher.RunOnDispatcherThreadAsync(() =>
+            {
+                _documentResolver.TryResolveDocument(
+                    request.TextDocument.Uri.GetAbsoluteOrUNCPath(), out var documentSnapshot);
+
+                return documentSnapshot;
+            }, cancellationToken);
+
+            if (document is null || cancellationToken.IsCancellationRequested)
+            {
+                return null;
+            }
+
+            var codeDocument = await document.GetGeneratedOutputAsync();
+            if (codeDocument.IsUnsupported())
+            {
+                return null;
+            }
+
+            var location = await GetSourceLocation(request, document).ConfigureAwait(false);
+            var ancestors = GetAncestors(codeDocument, location);
```

Comment:
Interesting, why capture every ancestor versus just looking for the first ancestor that's an html/taghelper element? I worry that this may be too heavy handed since this endpoint is called on typing paths